OpenRA.Mods.Sp/Traits/World/CloudSpawner.cs

```diff
@@ -1,99 +1,182 @@
-		[Desc("Cloud forward movement. Two values mean the cloud speed randomizes between them.")]
-		public readonly WDist[] Speed;
-
-		[Desc("The altitude of the cloud.")]
-		public readonly WDist CruiseAltitude = new WDist(2560);
-
-		[Desc("Distance margin where the cloud can be removed.")]
-		public readonly WDist CloseEnough = new WDist(128);
-
-		public override object Create(ActorInitializer init) { return new CloudSpawner(this); }
-	}
-
-	public class CloudSpawner : ITick
-	{
-		readonly CloudSpawnerInfo info;
-
-		int ticks;
-
-		public CloudSpawner(CloudSpawnerInfo info)
-		{
-			this.info = info;
-
-			ticks = info.InitialSpawnDelay;
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Sp.Effects;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Sp.Traits
+{
+	public class CloudSpawnerInfo : TraitInfo, ILobbyCustomRulesIgnore
+	{
+		[Desc("Average time (ticks) between cloud spawn.")]
+		public readonly int SpawnInterval = 10 * 25;
+
+		[Desc("Delay (in ticks) before the first cloud spawns.")]
+		public readonly int InitialSpawnDelay = 0;
+
+		[FieldLoader.Require]
+		[Desc("Which image to use.")]
+		public readonly string Image;
+
+		[FieldLoader.Require]
+		[Desc("Which sequence to use.")]
+		[SequenceReference("Image")]
+		public readonly string[] Sequences;
+
+		[FieldLoader.Require]
+		[Desc("Which palette to use.")]
+		[PaletteReference]
+		public readonly string Palette;
+
+		[Desc("Facing that the cloud may approach from.")]
+		public readonly int WindDirection = 8;
+
+		[Desc("Spawn and remove the cloud this far outside the map.")]
+		public readonly WDist Cordon = new WDist(7680);
+
+		[FieldLoader.Require]
+		[Desc("Cloud forward movement. Two values mean the cloud speed randomizes between them.")]
+		public readonly WDist[] Speed;
+
+		[Desc("The altitude of the cloud.")]
+		public readonly WDist CruiseAltitude = new WDist(2560);
+
+		[Desc("Distance margin where the cloud can be removed.")]
+		public readonly WDist CloseEnough = new WDist(128);
+
+		[Desc("Should we pre-spawn clouds covers the map?")]
+		public readonly bool ShouldPrespawn = true;
+
+		public override object Create(ActorInitializer init) { return new CloudSpawner(this); }
+	}
+
+	public class CloudSpawner : ITick
+	{
+		readonly CloudSpawnerInfo info;
+
+		int ticks;
+		bool prespawn = false;
+
+		public CloudSpawner(CloudSpawnerInfo info)
+		{
+			this.info = info;
+			prespawn = info.ShouldPrespawn;
+			ticks = info.InitialSpawnDelay;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (prespawn)

```

Comment:
Can't you use `IWorldLoaded.WorldLoaded`?

Reply:
Thanks for mention this and I will try it.